far/global.hpp

```diff
@@ -42,14 +42,58 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "platform.hpp"
 
 // Common:
+#include "common/bytes_view.hpp"
 #include "common/preprocessor.hpp"
 
 // External:
 
 //----------------------------------------------------------------------------
 
-enum class lng;
-enum class search_case_fold;
+struct global_search_replace_params
+{
+	NONCOPYABLE(global_search_replace_params);
+	NONMOVABLE(global_search_replace_params);
+
+	global_search_replace_params() = default;
+
+						// It is client's responsibility to maintain invariants:
+	string SearchStr;   // If Hex, should be BytesToString(SearchBytes, CodePage)
+	bytes SearchBytes;  // If not Hex, should be empty
```

Comment:
:sigh: This will incur dependency of Global on code page and strmix. Even when I introduced this dependency some time ago, and in global.cpp only, I thought it was wrong, but I deferred cleaning up to this change. Now, with the templated function, the dependency would be in the header, which is even worse. Besides, there is no way to pass code page to the operator, but this is minor, of course.

On the other hand, the clients need this invariant for their own sanity, and they have all necessary dependencies anyway so not a big problem for them.

The philosophy is that the global params are just storage, they have no behavior. It is the clients that cooperate among themselves to maintain proper state. The `operator<<` is provided for convenience only, to reduce copy-paste of making repeating assignments on clients' sides.

If you think the dependency is acceptable, let me know, I'll add this logic here.

Reply:
Ah I see. Indeed, it is a hairy problem.

Conceptually, global.?pp is a workaround to address issues with initialization order of complex core objects, not a home for each and every global/static variable in the project.
Ideally we should tear it down, not build it up.
`global_search_replace_params` does not really belong here and, since it consists of simple vocabulary types, it does not need a special treatment.
We can probably move it to stddlg.?pp and do the usual thing there, e.g.:
```C++
global_search_replace_params& params()
{
    static global_search_replace_params s_Params;
    return s_Params;
}
```
It should also be possible to move the hex logic into a non-templated helper function to not pollute the header.

Reply:
> Conceptually, global.?pp is a workaround...

That changes everything! I reworked the whole as you suggested (plus a separate instance for help). Please see if you like it.